atomix/cluster/src/main/java/io/atomix/cluster/messaging/impl/DefaultClusterEventService.java

```diff
@@ -354,77 +202,64 @@ public boolean isRunning() {
-    }
-
-    /**
-     * Returns the message type.
-     *
-     * @return the message type
-     */
-    public Type type() {
-      return type;
-    }
-
-    /**
-     * Returns the payload.
-     *
-     * @return the payload
-     */
-    public byte[] payload() {
-      return payload;
-    }
-
-    private enum Type {
-      DIRECT,
-      ALL,
-    }
+  @Override
+  public void event(final ClusterMembershipEvent event) {
+    eventServiceExecutor.execute(
+        () -> {
+          switch (event.type()) {
+            case MEMBER_ADDED:
+              updateRemoteSubscription(event);
+              break;
+            case METADATA_CHANGED:
+              updateRemoteSubscription(event);
+              break;
+            case REACHABILITY_CHANGED:
+              break;
+            case MEMBER_REMOVED:
+              removeAllSubscription(event.subject().id());
+              break;
+            default:
+              LOGGER.warn(
+                  "Unexpected membership event type {} from {}", event.type(), event.subject());
+          }
+        });
   }
 
-  /** Subscriber iterator that iterates subscribers in a loop. */
-  private static class TopicIterator implements Iterator<InternalSubscriptionInfo> {
-    private final AtomicInteger counter = new AtomicInteger();
-    private final InternalSubscriptionInfo[] subscribers;
-
-    TopicIterator(final List<InternalSubscriptionInfo> subscribers) {
-      final List<InternalSubscriptionInfo> filteredSubscribers =
-          subscribers.stream().filter(s -> !s.isTombstone()).collect(Collectors.toList());
-      Collections.reverse(filteredSubscribers);
-      this.subscribers =
-          filteredSubscribers.toArray(new InternalSubscriptionInfo[filteredSubscribers.size()]);
-    }
-
-    @Override
-    public boolean hasNext() {
-      return subscribers.length > 0;
+  private void removeAllSubscription(final MemberId id) {
+    final Set<String> prevSubscriptions = remoteMemberSubscriptions.remove(id);
+    if (prevSubscriptions != null) {
+      prevSubscriptions.forEach(s -> topics.get(s).removeRemoteSubscription(id));
     }
+  }
 
-    @Override
-    public InternalSubscriptionInfo next() {
-      return subscribers[Math.abs(counter.incrementAndGet() % subscribers.length)];
+  private void updateRemoteSubscription(final ClusterMembershipEvent event) {
+    final String topicSubscribedAsString =
+        event.subject().properties().getProperty(SUBSCRIPTION_PROPERTY_NAME);
+    if (topicSubscribedAsString != null) {
+      final Set<String> topicsSubscribed = topicsFromString(topicSubscribedAsString);
+      topicsSubscribed.forEach(
+          topic ->
+              topics
+                  .computeIfAbsent(topic, t -> new InternalTopic(topic))
+                  .addRemoteSubscription(event.subject().id()));
+      remoteMemberSubscriptions.put(event.subject().id(), topicsSubscribed);
+    } else {
+      removeAllSubscription(event.subject().id());
```

Comment:
These properties are always send via gossip? Can it happen that we have an different update like our broker address or something similar, which doesn't contain the subscription name and then we remove all subscriptions here?

Reply:
I don't think it happens since Swim gossip versions the update and guarantees that we always see the latest one even if it is received out of order. Other option is to add tombstones to deleted subscriptions and purge them periodically. But I think it is unnecessary here.

Reply:
Also gossip update contains the whole Property. It wouldn't propagate only address or only topology info.

Reply:
The question was more do we send always all details? If so then it is probably fine.

Reply:
Yes. It sends the whole member Properties.